### PR TITLE
Overhaul of TestHelper and SecureMul 

### PR DESCRIPTION
**Pull request by akoshelev**
## Overview

Following up the offline conversation with @martinthomson, the test
framework, secure multiplication have been completely rewritten to
follow the new network layer API. See the main motivation in https://github.com/martinthomson/raw-ipa/pull/80,
this PR tries to finalize the shape of `QueryId`, `RecordId` and
`Step` as well as their use at the borderline between protocols and
infra. All concepts presented https://github.com/martinthomson/raw-ipa/pull/80 largely remain the same.

## List of changes compared to #80 

* Test helper network now uses bi-directional channels (implemented as two physical channels) per **query id** and **step** with a handshake protocol to establish those connections
* QueryId is just a unit struct for the moment as we don't have a use case for it.
* Multiplication is now available via `context` and can be requested by providing record id and step. The reason for it is to be able to implement backpressure later
* `HelperAddr` is dropped in favor of `Identity` 
*  Streaming multiply test has been moved to  the `stream` mod

### CODE REVIEW

File: src/helpers/mock.rs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+        loop {
+            // Depending on connection status, request a new connection, spin and wait for
+            // connection acknowledgment from peer or return the sender end of connection
+            // if it is ready
+            let control_message = {
+                let mut connections = self.connections.lock().unwrap();
+                let conn_state = connections
+                    .entry((peer, step))
+                    .or_insert_with(|| ConnectionState::Listen);
```
Comment (thurstonsand) [inline src/helpers/mock.rs:206]:
```suggestion
                    .or_insert(ConnectionState::Listen);
```

File: src/helpers/mesh.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    /// Receive a message that is associated with the given record id.
+    async fn receive<T: Message>(&mut self, source: Identity, record: RecordId)
+        -> Result<T, Error>;
+
+    /// Returns the unique identity of this helper.
+    fn identity(&self) -> Identity;
+}
+
+/// This is the entry point for protocols to request communication when they require it.
+pub trait Gateway<M: Mesh, S: Step> {
```
Comment (thurstonsand) [inline src/helpers/mesh.rs:44]: so to understand the workflow here --

protocol writers will have a long-lived `Gateway` they hold onto, and for each `Step` they take, they use the `Gateway` to generate a `Mesh` with which to send/receive? and this `Mesh` will be valid for the duration of a `Step`?

File: src/helpers/mock.rs [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+                        Some(ControlMessage::ConnectionRequest(self.identity, step, rx))
+                    }
+                    ConnectionState::Established(sender) => {
+                        return sender.clone();
+                    }
+                }
+            };
+
+            if let Some(msg) = control_message {
```
Comment (thurstonsand) [inline src/helpers/mock.rs:221]: is this not always a `Some(msg)`?

File: src/helpers/mock.rs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        // spin and wait until message with the same record id appears in the buffer
+        // when it happens, pop it out, try to reinterpret its bytes as `T` and return
+        loop {
+            {
+                let mut buf = self.buf.lock().unwrap();
+                if let Some(msgs) = buf.get_mut(&(peer, step)) {
+                    let l = msgs.len();
+                    for i in 0..l {
+                        if msgs[i].record_id == record_id {
+                            msgs.swap(i, l - 1);
```
Comment (thurstonsand) [inline src/helpers/mock.rs:243]: this works, but seems like it'd be easier to understand as a `HashMap<RecordId, MessageEnvelope>`

File: src/helpers/mod.rs [hunk 5]
```diff
@@ -1,11 +1,58 @@
...
+
+#[derive(Debug, Copy, Clone)]
+pub enum Direction {
+    Left,
+    Right,
+}
+
+impl Identity {
+    #[must_use]
+    pub fn all_variants() -> &'static [Identity; 3] {
```
Comment (thurstonsand) [inline src/helpers/mod.rs:23]: what's the difference between this and something like

```rust
const all_variants: &'static [Identity; 3] = &[Identity::H1, Identity::H2, Identity::H3];
```

File: src/lib.rs [hunk 6]
```diff
@@ -1,12 +1,15 @@
 #![deny(clippy::clone_on_ref_ptr)]
 
+extern crate core;
```
Comment (thurstonsand) [inline src/lib.rs:3]: what does this line do?

File: src/helpers/mock.rs (hunk 4 above)
Comment (akoshelev) [reply src/helpers/mock.rs:243]: yea, initially I thought that I should just have a linear buffer of objects but ended up implementing a lookup by `record_id` which is what hashmap is for. I'll probably keep it around as there are other ideas circulating about the internals so this method may be changed entirely

File: src/helpers/mock.rs (hunk 1 above)
Comment (akoshelev) [reply src/helpers/mock.rs]: Does anybody know why this suggestion is not visible from the files tab??

Comment (thurstonsand) [reply src/helpers/mock.rs]: if you've made a change to that line, it'll only show up on commits before the change. so you'd have to pick an earlier commit in the files tab to see it

File: src/helpers/mod.rs (hunk 5 above)
Comment (thurstonsand) [reply src/helpers/mod.rs:23]: oh you know i had no idea godbolt supported rust. cool!

File: src/helpers/mock.rs (hunk 4 above)
Comment (thurstonsand) [reply src/helpers/mock.rs:241]: 👍

File: src/helpers/mod.rs (hunk 5 above)
Comment (thurstonsand) [reply src/helpers/mod.rs:23]: you know what, there actually is a difference. there isn't a way to reference `const all_variants` without a reference to a particular instance, i.e. `self.all_variants`. while the `fn all_variants()` can be referenced via `Identity::all_variants()`

Comment (akoshelev) [reply src/helpers/mod.rs:23]: hmm, does the following cover this use case? 

```rust
impl Identity {
    pub const ALL_VARIANTS: &'static [Identity; 3] = &[Identity::H1, Identity::H2, Identity::H3];
   
    pub fn l(&self) -> usize {
        Self::ALL_VARIANTS.len()
    }
}

fn l() -> usize {
    Identity::ALL_VARIANTS.len()
}


```

**Comment by martinthomson** [comment]
I realize that we really need to get this moving.

File: src/helpers/mod.rs (hunk 5 above)
Comment (thurstonsand) [reply src/helpers/mod.rs:23]: oh, yeah. so it does work. my IDE was telling me that was invalid, but i didn't actually test it on compile. thanks CLion

**Pull request merged**